File: resources/solve6.m

% solve the minimum time path problem
% use sinusoidal parameterization to order n
n=4;

% read in the velocity data array defined on 
% [0,1]x[0,1] and set the path end points
pathpar=[];
pathpar.v=readmatrix('SpeedData.csv');
[my,mx]=size(pathpar.v);
pathpar.A=[.05 .05];
pathpar.B=[.95 .95];

% add obstruction
%[yo,xo]=ndgrid(1:my,1:mx);
%yc=my/2+55;xc=mx/2+45;
%pathpar.v((yo-yc).^2+(xo-xc).^2<1600)=0.01;

% set optimization parameters
pr.objective=@pathtime;
pr.par=pathpar;
pr.x0=0.1*randn(2*n,1);  
pr.method='BFGS';
pr.linesearch='StrongWolfe';
pr.dftol=1E-8;
pr.ngtol=1E-8;
pr.dxtol=1E-8;
pr.c1=0.001;
pr.c2=0.9;
pr.m=5;
pr.maxiter=999;
pr.progress=10;

% call the optimization routine
out=optimize(pr);

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% plot the optimal path overlayed on the velocity map
% first, set up graphics parameters
FontSize=24;        % size of figure fonts
LineWidth=3;        % width of path line
PointSize=100;      % path end point area
LineColor=[1 1 1];  % path color
ColorMap=jet(512);  % velocity color map
NumPathPoints=1000; % number of points (s) defining the path
FigureScale=1.4;    % figure size scale on screen

% start the figure drawing
figure('position',FigureScale*[200 200 950 800]);
% draw the velocity map
imagesc(pathpar.v)
colormap(ColorMap)
% set axes parameters
set(gca,'xtick',[],'ytick',[],'box','on')
hc=colorbar('fontsize',FontSize);
TL=get(hc,'ticklabels');
for k=1:length(TL)
    TL{k}=num2str(str2double(TL{k}),'%.1f');
end
set(hc,'ticklabels',TL);
% and add the starting and ending positions to the plot
hold on
scatter([pathpar.A(1) pathpar.B(1)]*mx,[pathpar.A(2) pathpar.B(2)]*my,...
    PointSize,LineColor,'filled')
% compute the path and add to the plot (this must match
% the computation of the objective function)
s=linspace(0,1,NumPathPoints)';
xx=(1-s)*pathpar.A(1)+s*pathpar.B(1);
yy=(1-s)*pathpar.A(2)+s*pathpar.B(2);
for k=1:n
    S=sin(k*pi*s);
    xx=xx+out.x(k,end)*S;
    yy=yy+out.x(k+n,end)*S;
end
xxs=1+xx*(mx-1);
yys=1+yy*(my-1);
plot(xxs,yys,'color',LineColor,'linewidth',LineWidth)

%saveas(gcf,'minpathex.png')

% start the figure drawing
figure('position',FigureScale*[200 200 950 800]);
% draw the velocity map
imagesc(pathpar.v)
colormap(ColorMap)
% set axes parameters
set(gca,'xtick',[],'ytick',[],'box','on')
hc=colorbar('fontsize',FontSize);
TL=get(hc,'ticklabels');
for k=1:length(TL)
    TL{k}=num2str(str2double(TL{k}),'%.1f');
end
set(hc,'ticklabels',TL);
% and add the starting and ending positions to the plot
hold on
scatter([pathpar.A(1) pathpar.B(1)]*mx,[pathpar.A(2) pathpar.B(2)]*my,...
    PointSize,LineColor,'filled')

File: resources/solve6.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
script for solving the Minimium Time Path Problem
Version: November 9, 2023
Author: Tom Asaki
"""

import numpy as np
import optimize as opt
import pandas as pd
import matplotlib.pyplot as plt

from objective import pathtime as obj

N=12  # This is the order of the fit (2N decision variables)

# read in the velocity data array defined on 
# [0,1]x[0,1] and set the path end points
v=pd.read_csv('SpeedData.csv',header=None).to_numpy()
my,mx=v.shape
A=(.05,.05)
B=(.95,.95)

alg=dict(obj        = obj,
         x0         = 0.2*np.random.randn(2*N,1),
         params     = (v,A,B),
         method     = 'BFGS',
         maxiter    = 999,
         progress   = 10,
         ngtol      = 1E-8,
         dftol      = 1E-8,
         dxtol      = 1E-8,
         Lambda     = 1.,
         Lambdamax  = 100.,
         linesearch = 'StrongWolfe',
         c1         = 0.001,
         c2         = 0.9,
         m          = 10,
         maxcond    = 1000,
         )   

res=opt.minimize(alg)

######################################################################
# plot the optimal path superimposed on the velocity image
smp         = 1000      # number of points defining the path
FigDPI      = 256       # figure dpi (effects scale)
FigSize     = (8,6)     # fiugure size   
ColorMap    = 'jet'     # velocity colormap
LineColor   = 'white'   # path plot color
LineWidth   = 1         # path line width
PointSize   = 16        # size of path endpoints

r=np.linspace(0,1,smp)
xx=(1-r)*A[0]+r*B[0]
yy=(1-r)*A[1]+r*B[1]
for k in range(N):
    s=np.sin((k+1)*np.pi*r)
    xx+=res['x'][k,-1]*s
    yy+=res['x'][k+N,-1]*s
xxr=xx*(mx-1)
yyr=yy*(my-1)

fig = plt.figure(dpi=FigDPI,figsize=FigSize)
ax=fig.add_subplot()
vim=ax.imshow(v,cmap=ColorMap)
plt.colorbar(vim,orientation='vertical')
ax.plot(yyr,xxr,color=LineColor,linewidth=LineWidth)
ax.scatter(A[1]*mx,A[0]*my,PointSize,LineColor)
ax.scatter(B[1]*mx,B[0]*my,PointSize,LineColor)
plt.xticks([])
plt.yticks([])
plt.show()



